base/notify.go

```diff
@@ -0,0 +1,247 @@
+		EscapedJSONStringReport: escapedReport,
+	}, nil
+}
+
+// getPayload fetches the payload template from the PayloadTemplateURL and
+// executes it with values from getPayloadTemplateValues()
+func (t *notifyTask) getPayload(exp *Experiment) (string, error) {
+	if t.With.PayloadTemplateURL != "" {
+		template, err := getProviderTemplate(t.With.PayloadTemplateURL)
+		if err != nil {
+			log.Logger.Error("could not get payload template")
+			return "", err
+		}
+
+		values, err := getPayloadTemplateValues(exp)
+		if err != nil {
+			log.Logger.Error("could not get payload template values")
+			return "", err
+		}
+
+		// get the metrics spec
+		var buf bytes.Buffer
+		err = template.Execute(&buf, values)
+		if err != nil {
+			log.Logger.Error("could not execute payload template")
+			return "", err
+		}
+
+		return buf.String(), nil
+	}
+
+	return "", nil
+}
+
+// initializeDefaults sets default values for the custom metrics task
+func (t *notifyTask) initializeDefaults() {
+	// set default HTTP method
+	if t.With.Method == "" {
+		if t.With.PayloadTemplateURL != "" {
+			t.With.Method = "POST"
+		} else {
+			t.With.Method = "GET"
+		}
+	}
+}
+
+// validate task inputs
+func (t *notifyTask) validateInputs() error {
+	return nil
+}
+
+// run executes this task
+func (t *notifyTask) run(exp *Experiment) error {
+	// validate inputs
+	err := t.validateInputs()
+	if err != nil {
+		return err
+	}
+
+	if exp.driver == nil {
+		return errors.New("no driver was provided for collect-metrics-database task")
+	}
+
+	// initialize defaults
+	t.initializeDefaults()
+
+	var requestBody io.Reader
+
+	log.Logger.Debug("method: ", t.With.Method, " URL: ", t.With.Url)
+
+	if t.With.PayloadTemplateURL != "" {
+		payload, err := t.getPayload(exp)
+		if err != nil {
+			log.Logger.Error("could not get payload")
+			return err
+		}
+
+		log.Logger.Debug("add payload: ", string(payload))
+
+		requestBody = strings.NewReader(payload)
+	}
+
+	// create a new HTTP request
+	req, err := http.NewRequest(t.With.Method, t.With.Url, requestBody)
+	if err != nil {
+		log.Logger.Error("could not create HTTP request for notify task:", err)
+		return nil
+	}
+
+	// iterate through headers
+	for headerName, headerValue := range t.With.Headers {
+		req.Header.Add(headerName, headerValue)
+		log.Logger.Debug("add header: ", headerName, ", value: ", headerValue)
+	}
+
+	// add query params
+	q := req.URL.Query()
+	for key, value := range t.With.Params {
+		q.Add(key, value)
+		log.Logger.Debug("add param: ", key, ", value: ", value)
+	}
+	req.URL.RawQuery = q.Encode()
+
+	// send request
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Logger.Error("could not send HTTP request for notify task:", err)
+		return nil
+	}
+	defer resp.Body.Close()
+
+	// read response responseBody
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		log.Logger.Error("could not read response body from Slack notification request", err)
+		return nil
+	}
```

Comment:
should we also be looking at resp.StatusCode() and failing if not 200/202 ?

Reply:
Yeah, I think that is the point of `softFailure` which I forgot to implement. I will do that.